File: desktop/macos/Desktop/Sources/Audio/BleAudioService.swift
import Combine
import Foundation
import os.log

// MARK: - BLE Audio Service

/// Service that coordinates BLE device audio processing and transcription
/// Connects device connections to the transcription pipeline
@MainActor
final class BleAudioService: ObservableObject {

  // MARK: - Singleton

  static let shared = BleAudioService()

  // MARK: - Published Properties

  @Published private(set) var isProcessing = false
  @Published private(set) var currentCodec: BleAudioCodec?
  @Published private(set) var audioLevel: Float = 0.0
  @Published private(set) var isDecodeDegraded = false

  // MARK: - Properties

  private let logger = Logger(subsystem: "me.omi.desktop", category: "BleAudioService")
  private var processor: BleAudioProcessor?
  private var audioStreamTask: Task<Void, Never>?
  private var cancellables = Set<AnyCancellable>()

  // Audio delivery
  private var transcriptionService: TranscriptionService?
  private var audioDataHandler: ((Data) -> Void)?
  private var rawFrameHandler: ((Data) -> Void)?

  // Statistics
  private var totalSamplesProcessed: Int = 0
  private var startTime: Date?

  // MARK: - Initialization

  private init() {}

  // MARK: - Public Methods

  /// Start processing audio from a device connection
  /// - Parameters:
  ///   - connection: The device connection to get audio from
  ///   - transcriptionService: Optional transcription service to send audio to
  ///   - audioDataHandler: Optional handler for decoded PCM data (alternative to transcription)
  ///   - rawFrameHandler: Optional handler for raw encoded frames (for WAL recording)
  func startProcessing(
    from connection: DeviceConnection,
    transcriptionService: TranscriptionService? = nil,
    audioDataHandler: ((Data) -> Void)? = nil,
    rawFrameHandler: ((Data) -> Void)? = nil
  ) async {
    guard !isProcessing else {
      logger.warning("Already processing audio")
      return
    }
    // Claim the slot synchronously, before the first await, so two overlapping
    // startProcessing calls cannot both pass the guard and double-create the
    // processor (the second would orphan the first's processor + stream task).
    isProcessing = true

    self.transcriptionService = transcriptionService
    self.audioDataHandler = audioDataHandler
    self.rawFrameHandler = rawFrameHandler

    // Get codec from device
    let codec = await connection.getAudioCodec()
    currentCodec = codec

    // Check if codec is supported
    if !AudioDecoderFactory.isSupported(codec) {
      logger.error("Unsupported audio codec: \(codec.name)")
      // Release the claimed slot and drop the handlers captured above.
      stopProcessing()
      return
    }

    // Warn if codec has partial support
    if !AudioDecoderFactory.hasFullSupport(codec) {
      logger.warning("Codec \(codec.name) has partial support - audio quality may be affected")
    }

    // Create processor
    processor = BleAudioProcessor(codec: codec)
    processor?.delegate = self
    isDecodeDegraded = false

    // Subscribe to decoded PCM data
    processor?.pcmDataPublisher
      .receive(on: DispatchQueue.main)
      .sink { [weak self] pcmData in
        self?.handleDecodedAudio(pcmData)
      }
      .store(in: &cancellables)

    // Start audio stream from device
    let audioStream = connection.getAudioStream()

    isProcessing = true
    startTime = Date()
    totalSamplesProcessed = 0

    logger.info("Started processing audio with codec: \(codec.name)")

    // Process audio stream
    audioStreamTask = Task { [weak self] in
      do {
        for try await audioData in audioStream {
          guard let self = self, self.isProcessing else { break }

          // Process based on device type
          await self.processDeviceAudio(audioData, from: connection)
        }
      } catch {
        self?.logger.error("Audio stream error: \(error.localizedDescription)")
      }

      // The stream ended or errored. Run full cleanup (not just isProcessing =
      // false), otherwise the processor, Combine subscriptions, and handlers
      // dangle and the session cannot cleanly restart.
      await self?.handleAudioStreamEnded()
    }
  }

  /// Full teardown after the device audio stream ends or errors on its own.
  private func handleAudioStreamEnded() {
    guard isProcessing else { return }
    logger.info("Audio stream ended; tearing down processing")
    stopProcessing()
  }

  /// Stop processing audio. Idempotent: safe to call after the stream has
  /// already ended (the old `guard isProcessing` early-return skipped cleanup
  /// in exactly that case, leaving the session unrecoverable).
  func stopProcessing() {
    audioStreamTask?.cancel()
    audioStreamTask = nil
    processor?.reset()
    cancellables.removeAll()

    isProcessing = false
    transcriptionService = nil
    audioDataHandler = nil
    rawFrameHandler = nil

    // Log statistics
    if let start = startTime {
      let duration = Date().timeIntervalSince(start)
      let stats = processor?.getStatistics() ?? (frames: 0, bytes: 0, lostPackets: 0)
      logger.info(
        "Stopped processing. Duration: \(String(format: "%.1f", duration))s, Frames: \(stats.frames), Bytes: \(stats.bytes), Lost: \(stats.lostPackets)"
      )
    }

    startTime = nil
    processor = nil
    currentCodec = nil
  }

  // MARK: - Private Methods

  /// Process audio data from a device
  private func processDeviceAudio(_ data: Data, from connection: DeviceConnection) async {
    guard let processor = processor else { return }

    // Capture raw frame for WAL recording
    rawFrameHandler?(data)

    // Different devices need different handling
    let deviceType = connection.device.type

    switch deviceType {
    case .fieldy:
      // Fieldy sends pre-framed 40-byte Opus frames
      processor.processAudioData(data)

    case .friendPendant:
      // Friend Pendant sends 30-byte LC3 frames (already extracted by connection)
      processor.processAudioData(data)

    case .bee:
      // Bee sends ADTS-framed AAC (already parsed by connection)
      processor.processFrame(data)

    case .limitless:
      // Limitless sends Opus frames extracted from protobuf
      processor.processFrame(data)

    case .plaud:
      // PLAUD sends chunked Opus data
      processor.processAudioData(data)

    case .omi, .openglass:
      // Omi devices send packet-framed audio
      processor.processAudioData(data)

    default:
      // Default: treat as raw frames
      processor.processAudioData(data)
    }
  }

  /// Handle decoded PCM audio
  private func handleDecodedAudio(_ pcmData: Data) {
    guard !pcmData.isEmpty else { return }

    totalSamplesProcessed += pcmData.count / 2

    // Calculate audio level
    updateAudioLevel(from: pcmData)

    // Send to transcription service (mono — Python backend handles diarization server-side)
    if let transcription = transcriptionService {
      transcription.sendAudio(pcmData)
    }

    // Send to custom handler
    audioDataHandler?(pcmData)
  }

  /// Convert mono PCM to stereo (duplicate to both channels)
  private func convertToStereo(_ monoData: Data) -> Data {
    // Mono: [S0, S1, S2, ...]
    // Stereo: [S0, S0, S1, S1, S2, S2, ...] (interleaved)
    var stereoData = Data(capacity: monoData.count * 2)

    monoData.withUnsafeBytes { bytes in
      let samples = bytes.bindMemory(to: Int16.self)
      for i in 0..<samples.count {
        var sample = samples[i]
        // Write same sample to both channels
        stereoData.append(Data(bytes: &sample, count: 2))
        stereoData.append(Data(bytes: &sample, count: 2))
      }
    }

    return stereoData
  }

  /// Calculate RMS audio level from PCM data
  private func updateAudioLevel(from data: Data) {
    var sumSquares: Float = 0
    let sampleCount = data.count / 2

    data.withUnsafeBytes { bytes in
      let samples = bytes.bindMemory(to: Int16.self)
      for i in 0..<samples.count {
        let sample = Float(samples[i]) / 32768.0
        sumSquares += sample * sample
      }
    }

    let rms = sqrt(sumSquares / Float(max(sampleCount, 1)))
    // Smooth the level
    audioLevel = audioLevel * 0.7 + rms * 0.3
  }
}

// MARK: - Convenience Extensions

extension BleAudioService {

  /// Check if a device's audio codec is supported
  func isCodecSupported(for connection: DeviceConnection) async -> Bool {
    let codec = await connection.getAudioCodec()
    return AudioDecoderFactory.isSupported(codec)
  }

  /// Get codec information for a device
  func getCodecInfo(for connection: DeviceConnection) async -> (codec: BleAudioCodec, supported: Bool, name: String) {
    let codec = await connection.getAudioCodec()
    let supported = AudioDecoderFactory.isSupported(codec)
    return (codec, supported, codec.name)
  }
}

// MARK: - BleAudioProcessorDelegate

extension BleAudioService: BleAudioProcessor.Delegate {
  nonisolated func bleAudioProcessor(_ processor: BleAudioProcessor, didDecodeSamples samples: [Int16]) {
    // PCM delivery uses pcmDataPublisher; delegate path is unused.
    // Reset the degraded flag on successful decode so it reflects the
    // current processor state rather than staying sticky.
    Task { @MainActor [weak self] in
      guard let self, self.isDecodeDegraded else { return }
      self.isDecodeDegraded = false
      self.logger.info("BLE decode recovered — clearing degraded flag")
    }
  }

  nonisolated func bleAudioProcessor(_ processor: BleAudioProcessor, didFailWithError error: Error) {
    Task { @MainActor [weak self] in
      guard let self else { return }
      self.isDecodeDegraded = true
      self.logger.error("BLE decode degraded: \(error.localizedDescription)")
    }
  }
}

// MARK: - Integration with DeviceProvider

File: desktop/windows/src/renderer/src/device/audio/bleAudioService.ts
/**
 * BLE audio service — Windows port of macOS Audio/BleAudioService.swift. Owns
 * one processing session at a time: reads the device codec, builds the
 * processor, routes each family's frames to the right entry point, and hands
 * decoded 16 kHz mono PCM to the transcription sink.
 *
 * Every device family is decoded here, so the wire codec never reaches the
 * socket; the lane always carries linear16.
 */

import type { BleAudioCodec } from '../protocol/deviceTypes'
import type { DeviceConnection } from '../connections/deviceConnection'
import { BleAudioProcessor } from './bleAudioProcessor'
import { hasFullCodecSupport, isCodecSupported } from './audioDecoderFactory'

export interface BleAudioSessionHandlers {
  /** Decoded 16 kHz mono s16 PCM, in capture order. */
  onPcm: (pcm: Int16Array) => void
  /** The encoded device frame, for WAL capture, before any decoding. */
  onRawFrame?: (frame: Uint8Array) => void
  onCodec?: (codec: BleAudioCodec) => void
  onLevel?: (level: number) => void
  onDegradedChange?: (isDegraded: boolean) => void
  /** Fires when the device stream ends on its own (disconnect, teardown). */
  onEnded?: (error: Error | null) => void
}

export interface BleAudioSessionStats {
  framesProcessed: number
  bytesProcessed: number
  lostPackets: number
  decodeFailures: number
  durationMs: number
}

export class BleAudioService {
  private processing = false
  private generation = 0
  private processor: BleAudioProcessor | null = null
  private subscription: { cancel: () => void } | null = null
  private handlers: BleAudioSessionHandlers | null = null
  private startedAt = 0
  private codec: BleAudioCodec | null = null
  private level = 0
  private degraded = false

  get isProcessing(): boolean {
    return this.processing
  }

  get currentCodec(): BleAudioCodec | null {
    return this.codec
  }

  get audioLevel(): number {
    return this.level
  }

  get isDecodeDegraded(): boolean {
    return this.degraded
  }

  /**
   * Starts decoding the connection's audio. The processing slot is claimed
   * synchronously, before the first await, so two overlapping starts cannot
   * both build a processor.
   */
  async startProcessing(
    connection: DeviceConnection,
    handlers: BleAudioSessionHandlers
  ): Promise<boolean> {
    if (this.processing) {
      console.warn('[device] BLE audio processing already running')
      return false
    }
    this.processing = true
    const generation = ++this.generation
    this.handlers = handlers
    this.startedAt = Date.now()
    this.level = 0
    this.degraded = false

    let codec: BleAudioCodec
    try {
      codec = await connection.getAudioCodec()
    } catch (error) {
      // The slot is claimed synchronously, so a throw anywhere in setup has to
      // release it or every later start is refused as already running.
      console.error(`[device] failed to read the device audio codec: ${String(error)}`)
      this.stopProcessing()
      return false
    }
    // A stop can land during the codec read; that start is stale.
    if (!this.processing || this.generation !== generation) return false

    this.codec = codec
    handlers.onCodec?.(codec)

    if (!isCodecSupported(codec)) {
      console.error(`[device] unsupported audio codec ${codec}; not starting BLE audio`)
      this.stopProcessing()
      return false
    }
    if (!hasFullCodecSupport(codec)) {
      console.warn(`[device] ${codec} has partial support; audio quality may be affected`)
    }

    const processor = new BleAudioProcessor({
      codec,
      delegate: {
        onPcm: (pcm) => this.handleDecodedAudio(pcm, generation),
        onDegradedChange: (isDegraded) => {
          if (this.generation !== generation) return
          this.degraded = isDegraded
          handlers.onDegradedChange?.(isDegraded)
        }
      }
    })
    // No second decoder-existence check: createAudioDecoder covers the whole
    // codec union and returns null only for 'unknown', which the guard above
    // already rejected.
    try {
      await processor.ready()
    } catch (error) {
      console.error(`[device] audio decoder failed to initialize: ${String(error)}`)
      processor.close()
      this.stopProcessing()
      return false
    }
    if (!this.processing || this.generation !== generation) {
      processor.close()
      return false
    }
    this.processor = processor

    this.subscription = connection.getAudioStream({
      onValue: (frame) => {
        if (this.generation !== generation) return
        this.routeFrame(connection, frame)
      },
      onFinish: (error) => {
        if (this.generation !== generation) return
        // A stream that ends tears the whole session down, not just a flag.
        const ended = this.handlers?.onEnded
        this.stopProcessing()
        ended?.(error)
      }
    })
    return true
  }

  stopProcessing(): BleAudioSessionStats | null {
    if (!this.processing) return null
    this.processing = false
    this.generation += 1
    this.subscription?.cancel()
    this.subscription = null
    const processor = this.processor
    const stats = processor?.snapshot ?? null
    processor?.reset()
    processor?.close()
    this.processor = null
    this.handlers = null
    this.codec = null
    this.level = 0
    if (stats === null) return null
    const sessionStats: BleAudioSessionStats = {
      ...stats,
      durationMs: Date.now() - this.startedAt
    }
    console.info(
      `[device] BLE audio session ended: ${sessionStats.framesProcessed} frames, ` +
        `${sessionStats.bytesProcessed} bytes, ${sessionStats.lostPackets} lost packets, ` +
        `${sessionStats.decodeFailures} decode failures, ${sessionStats.durationMs} ms`
    )
    return sessionStats
  }

  /** Per-family frame routing: the connection layer already reassembles some
   *  families into whole frames, while others hand over raw notifications. */
  private routeFrame(connection: DeviceConnection, frame: Uint8Array): void {
    const processor = this.processor
    if (processor === null) return
    this.handlers?.onRawFrame?.(frame)
    switch (connection.device.type) {
      case 'bee':
      case 'limitless':
        processor.processFrame(frame)
        return
      default:
        processor.processAudioData(frame)
    }
  }

  private handleDecodedAudio(pcm: Int16Array, generation: number): void {
    if (this.generation !== generation) return
    this.updateAudioLevel(pcm)
    // Mono is deliberate: diarization happens server-side.
    this.handlers?.onPcm(pcm)
  }

  private updateAudioLevel(pcm: Int16Array): void {
    if (pcm.length === 0) return
    let sum = 0
    for (let i = 0; i < pcm.length; i += 1) {
      const sample = pcm[i] / 32768
      sum += sample * sample
    }
    const rms = Math.sqrt(sum / pcm.length)
    this.level = this.level * 0.7 + rms * 0.3
    this.handlers?.onLevel?.(this.level)
  }
}
